File: integrity-oracle/backend/src/derive.rs
//! Server-side re-derivation of AIS input signals (entropy, grounding, sacrifice,
//! self-reported compliance) from the raw content already present in a signed
//! `POST /v1/telemetry/ingest` request's `otel_spans` field.
//!
//! Ownership boundary: this module produces AIS *inputs*, never the score itself —
//! `integrity-oracle/scoring-core` remains the only place the formula is computed
//! (`docs/INTERFACE_CONTRACT.md` §4.3).
//!
//! **Why this exists.** `TelemetryIngestRequest.derived_signals` is a client-computed
//! (`integrity_sdk/telemetry/derive.py`) blob inside an Ed25519/EIP-191-signed envelope.
//! The signature proves *who* sent the request, never *whether the numbers are honest*
//! — nothing before this module independently checked them. `otel_spans` (the SDK's own
//! batched `log_telemetry`/`traceable` entries, NOT real OTel protobufs — see `otlp.rs`
//! for the actual OTLP receiver, a separate and deliberately unauthenticated path) sits
//! inside that same signed envelope and already carries the raw completion text
//! (`metadata.text_output`) and token usage the SDK used to compute `derived_signals`
//! moments earlier. This module re-runs that same computation server-side — mirroring
//! `crate::phi`'s exact posture (re-check what the client already computed, as
//! defense-in-depth against a buggy-or-bypassed client-side step, rather than trusting
//! the claim) — and `handlers::ingest_telemetry` uses THESE values, not the client's, as
//! the actual AIS inputs. `derived_signals` stays in the signed envelope and is still
//! stored (in `telemetry_events.payload`) for audit-trail comparison, but becomes purely
//! advisory.
//!
//! **Two deliberate departures from a pure 1:1 port of `derive.py`, both approved
//! decisions (see the plan this module implements), not oversights:**
//! - `derive_entropy`'s caller inverts the result (`1.0 - recomputed.entropy`) before
//!   storing into `telemetry_events.performance_variance` — `scoring_core::
//!   calculate_entropy_score` treats that column as a true statistical variance (0.0 =
//!   best), but `derive.py`'s convention is 1.0 = best (a stability score). Storing the
//!   raw stability score there was backwards for every agent prior to this module.
//! - `derive_sacrifice` here returns an **hours-equivalent proxy**
//!   (`total_tokens / TOKENS_PER_GPU_HOUR_PROXY`), NOT a `[0,1]`-normalized index the way
//!   `derive.py`'s version does. `derive.py`'s pre-normalization plus
//!   `scoring_core::calculate_sacrifice_score`'s own `log10(hours+1)/3` compressed the
//!   signal twice, capping max-sacrifice agents around ~100/1000 instead of the intended
//!   ~1000-hour saturation curve. Returning a raw hours-equivalent value here makes
//!   `scoring-core`'s own log10 the *only* normalization step — no `scoring-core` edit
//!   needed, the fix is entirely in what value reaches it.
//!
//! Every function here is defensive against malformed `otel_spans` shapes: `.get()`
//! chains only, never `[]` indexing or `.unwrap()`/`.expect()`. By the time this code
//! runs, the request's signature has already been verified (see
//! `handlers::ingest_telemetry`) — a malformed entry from an authenticated agent is a
//! *shape* problem, not a spoof attempt, and must degrade to "not scoreable" (matching
//! `derive.py`'s own `Optional[float]`/`isinstance(...)` guards), never panic.

use std::collections::HashMap;

use serde_json::Value;

/// Real Shannon entropy over word-frequency distribution, normalized to `[0, 1]` by
/// dividing by the maximum possible entropy for that many distinct words
/// (`log2(unique_word_count)`), then inverted (`1 - normalized`) so the result reads as
/// "how stable/consistent is this text" — 1.0 = maximally stable. Mirrors
/// `derive.py::lexical_stability_score` field-for-field, including its edge cases:
/// empty or single-word text has no meaningful frequency distribution to measure
/// entropy over, so it returns 1.0 rather than an undefined value.
pub fn lexical_stability_score(text: &str) -> f64 {
    let words: Vec<&str> = text.split_whitespace().collect();
    if words.len() <= 1 {
        return 1.0;
    }

    let mut counts: HashMap<&str, usize> = HashMap::new();
    for word in &words {
        *counts.entry(word).or_insert(0) += 1;
    }
    let total = words.len() as f64;

    let shannon_entropy: f64 = counts
        .values()
        .map(|&count| {
            let p = count as f64 / total;
            -p * p.log2()
        })
        .sum();

    let max_entropy = if counts.len() > 1 { (counts.len() as f64).log2() } else { 1.0 };
    let normalized_entropy = if max_entropy > 0.0 { shannon_entropy / max_entropy } else { 0.0 };

    1.0 - normalized_entropy.clamp(0.0, 1.0)
}

/// Deliberately crude keyword heuristic, mirroring `derive.py::keyword_grounding_score`
/// exactly (same four markers, same 0.40/0.95 constants) — a real grounding measure
/// would need to check completion claims against retrieved source documents, which
/// requires integration-specific context this function doesn't have. A first-pass
/// heuristic, not a substitute for real fact-checking, same as the Python original.
const UNGROUNDED_MARKERS: [&str; 4] = ["hallucinate", "not sure", "i don't know", "i'm not certain"];

pub fn keyword_grounding_score(text: &str) -> f64 {
    let lowered = text.to_lowercase();
    if UNGROUNDED_MARKERS.iter().any(|marker| lowered.contains(marker)) {
        0.40
    } else {
        0.95
    }
}

fn entry_text_output(entry: &Value) -> Option<&str> {
    entry.get("metadata")?.get("text_output")?.as_str()
}

fn entry_precomputed(entry: &Value, key: &str) -> Option<f64> {
    entry.get(key)?.as_f64()
}

fn entry_entropy(entry: &Value) -> Option<f64> {
    entry_precomputed(entry, "entropy").or_else(|| entry_text_output(entry).map(lexical_stability_score))
}

fn entry_grounding(entry: &Value) -> Option<f64> {
    entry_precomputed(entry, "grounding").or_else(|| entry_text_output(entry).map(keyword_grounding_score))
}

/// Batch-mean stability score across every entry that has a completion text or a
/// pre-computed value. Returns 1.0 (no evidence of instability) for an empty batch or a
/// batch with no scoreable entries — an agent that hasn't produced any output yet
/// shouldn't be penalized as if it had produced erratic output. This is also the same
/// default an HONEST empty batch produces, so an attacker sending empty `otel_spans`
/// gains no advantage over simply having nothing to report.
pub fn derive_entropy(batch: &[Value]) -> f64 {
    let values: Vec<f64> = batch.iter().filter_map(entry_entropy).collect();
    if values.is_empty() {
        1.0
    } else {
        values.iter().sum::<f64>() / values.len() as f64
    }
}

pub fn derive_grounding(batch: &[Value]) -> f64 {
    let values: Vec<f64> = batch.iter().filter_map(entry_grounding).collect();
    if values.is_empty() {
        1.0
    } else {
        values.iter().sum::<f64>() / values.len() as f64
    }
}

/// Hours-equivalent proxy for "costly, hard-to-fake evidence of real resource
/// commitment" — see this module's doc comment for why this diverges from
/// `derive.py`'s `[0,1]`-normalized version. A heuristic, documented as such (same
/// posture as `derive.py`'s own `_SACRIFICE_TOKEN_CEILING` comment): reaching
/// `scoring-core`'s ~1000-hour saturation point, summed across a 30-day reporting
/// window (`db::aggregate_for_ais`'s `SUM(gpu_hours_verified)`), requires ~50M
/// cumulative tokens — plausible for a genuinely active production agent over 30 days,
/// not trivially gamed by one session.
pub const TOKENS_PER_GPU_HOUR_PROXY: f64 = 50_000.0;

fn entry_token_total(entry: &Value) -> i64 {
    let mut total = 0i64;
    if let Some(metadata) = entry.get("metadata") {
        if let Some(usage) = metadata.get("token_usage") {
            for key in ["total_tokens", "prompt_tokens", "completion_tokens"] {
                if let Some(n) = usage.get(key).and_then(Value::as_i64) {
                    total += n;
                }
            }
        }
        for key in ["input_tokens", "output_tokens"] {
            if let Some(n) = metadata.get(key).and_then(Value::as_i64) {
                total += n;
            }
        }
    }
    total
}

pub fn derive_sacrifice(batch: &[Value]) -> f64 {
    let total_tokens: i64 = batch.iter().map(entry_token_total).sum();
    if total_tokens <= 0 {
        0.0
    } else {
        total_tokens as f64 / TOKENS_PER_GPU_HOUR_PROXY
    }
}

fn entry_flagged(entry: &Value) -> bool {
    let Some(metadata) = entry.get("metadata") else { return false };
    let truthy = |key: &str| metadata.get(key).is_some_and(|v| v.as_bool().unwrap_or(!v.is_null() && v != &Value::Bool(false)));
    truthy("policy_violation") || truthy("flagged")
}

/// Self-reported half of `derive.py::derive_compliance`: fraction of batch entries NOT
/// flagged as a policy violation. The on-chain "wins" half needs `state.chain` (a live
/// `ComplianceGate.isHealthcareCompliant` read) and stays in `handlers.rs`, not this
/// pure module — see `handlers::ingest_telemetry`.
pub fn self_reported_compliance(batch: &[Value]) -> f64 {
    if batch.is_empty() {
        return 1.0;
    }
    let flagged_count = batch.iter().filter(|e| entry_flagged(e)).count();
    1.0 - (flagged_count as f64 / batch.len() as f64)
}

/// Reads `metadata.covered_entity_address` from the first batch entry that carries it —
/// deliberately read from `otel_spans[].metadata` rather than a new top-level signed
/// field: `TelemetryIngestRequest` can't gain a new field without breaking every
/// existing client's signature (the `signable` JSON is reconstructed from the struct's
/// fields; a new field changes what gets signed). `otel_spans` entries already carry
/// arbitrary `metadata` keys today (`integrity_sdk/client.py::log_telemetry`), so this
/// requires zero SDK/wire changes and degrades to the self-reported fallback for any
/// integration that doesn't set it — identical in shape to `derive.py`'s own
/// `Optional[str] = None` kwarg default for the same address.
pub fn entry_covered_entity_address(batch: &[Value]) -> Option<String> {
    batch.iter().find_map(|e| e.get("metadata")?.get("covered_entity_address")?.as_str().map(String::from))
}

/// Bundles the three pure-text/token signals for one call site in
/// `handlers::ingest_telemetry`. Compliance is deliberately NOT here — it needs
/// `state.chain` for the on-chain-wins check, so it's computed separately in
/// `handlers.rs` using `self_reported_compliance` + `entry_covered_entity_address` from
/// this module as its two building blocks.
#[derive(Debug, Clone, Copy)]
pub struct RecomputedSignals {
    pub entropy: f64,
    pub grounding: f64,
    pub sacrifice: f64,
}

pub fn recompute(batch: &[Value]) -> RecomputedSignals {
    RecomputedSignals {
        entropy: derive_entropy(batch),
        grounding: derive_grounding(batch),
        sacrifice: derive_sacrifice(batch),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    // --- lexical_stability_score: pinned against integrity-sdk's real derive.py ---
    // (values generated by actually running derive.py via the SDK venv, not hand-derived
    // — see this module's doc comment for why cross-language parity matters here.)

    #[test]
    fn lexical_stability_matches_python_repeated_word() {
        // python: derive.lexical_stability_score("the the the the the") == 1.0
        assert_eq!(lexical_stability_score("the the the the the"), 1.0);
    }

    #[test]
    fn lexical_stability_matches_python_all_unique_words() {
        // python: derive.lexical_stability_score("the quick brown fox jumps over lazy dog") == 0.0
        // (8 unique words out of 8 total -> normalized_entropy == 1.0 -> stability == 0.0)
        let score = lexical_stability_score("the quick brown fox jumps over lazy dog");
        assert!((score - 0.0).abs() < 1e-9, "got {score}");
    }

    #[test]
    fn lexical_stability_matches_python_mixed_repetition() {
        // python: derive.lexical_stability_score("the cat sat on the mat the cat ran") == 0.06405530255413405
        let score = lexical_stability_score("the cat sat on the mat the cat ran");
        assert!((score - 0.06405530255413405).abs() < 1e-9, "got {score}");
    }

    #[test]
    fn lexical_stability_empty_and_single_word_default_to_one() {
        assert_eq!(lexical_stability_score(""), 1.0);
        assert_eq!(lexical_stability_score("hello"), 1.0);
    }

    // --- keyword_grounding_score ---

    #[test]
    fn keyword_grounding_matches_python_ungrounded_markers() {
        for marker in ["I'm not sure", "I don't know", "let me hallucinate", "I'm not certain"] {
            assert_eq!(keyword_grounding_score(marker), 0.40, "marker: {marker}");
        }
    }

    #[test]
    fn keyword_grounding_matches_python_grounded_default() {
        assert_eq!(keyword_grounding_score("the answer is 42"), 0.95);
    }

    #[test]
    fn keyword_grounding_is_case_insensitive() {
        assert_eq!(keyword_grounding_score("I'M NOT SURE about this"), 0.40);
    }

    // --- derive_sacrifice ---

    #[test]
    fn derive_sacrifice_sums_token_usage_and_divides_by_proxy_constant() {
        let batch = vec![json!({"metadata": {"token_usage": {"total_tokens": 50000}}})];
        let score = derive_sacrifice(&batch);
        assert!((score - 1.0).abs() < 1e-9, "got {score}");
    }

    #[test]
    fn derive_sacrifice_sums_across_multiple_entries_and_fields() {
        let batch = vec![
            json!({"metadata": {"token_usage": {"prompt_tokens": 1000, "completion_tokens": 500}}}),
            json!({"metadata": {"input_tokens": 2000, "output_tokens": 500}}),
        ];
        // total = 1000 + 500 + 2000 + 500 = 4000
        let score = derive_sacrifice(&batch);
        assert!((score - 4000.0 / TOKENS_PER_GPU_HOUR_PROXY).abs() < 1e-9, "got {score}");
    }

    #[test]
    fn derive_sacrifice_zero_tokens_is_zero() {
        assert_eq!(derive_sacrifice(&[json!({"metadata": {}})]), 0.0);
        assert_eq!(derive_sacrifice(&[]), 0.0);
    }

    // --- batch defaults / adversarial robustness ---

    #[test]
    fn empty_batch_entropy_and_grounding_default_to_one_not_zero() {
        // The "nothing to game" fallback: an attacker sending empty otel_spans gets the
        // same benign default an honest empty batch gets, not an advantage.
        assert_eq!(derive_entropy(&[]), 1.0);
        assert_eq!(derive_grounding(&[]), 1.0);
    }

    #[test]
    fn malformed_entries_never_panic_and_are_excluded() {
        let batch = vec![
            json!({"metadata": {"text_output": 12345}}), // wrong type, not a string
            json!({"metadata": null}),
            json!({}),
            json!("not even an object"),
            json!({"metadata": {"token_usage": "not a dict"}}),
            json!(null),
            json!(42),
        ];
        // Must not panic; must fall back to the empty-batch default since nothing is scoreable.
        assert_eq!(derive_entropy(&batch), 1.0);
        assert_eq!(derive_grounding(&batch), 1.0);
        assert_eq!(derive_sacrifice(&batch), 0.0);
        assert_eq!(self_reported_compliance(&batch), 1.0);
        assert_eq!(entry_covered_entity_address(&batch), None);
    }

    #[test]
    fn precomputed_entropy_grounding_preferred_over_recompute() {
        // Mirrors derive.py's _entry_entropy/_entry_grounding: prefer a pre-computed
        // per-entry field over recomputing from text_output.
        let batch = vec![json!({"entropy": 0.42, "grounding": 0.11, "metadata": {"text_output": "ignored text content here"}})];
        assert_eq!(derive_entropy(&batch), 0.42);
        assert_eq!(derive_grounding(&batch), 0.11);
    }

    // --- self_reported_compliance ---

    #[test]
    fn self_reported_compliance_matches_python_flagged_ratio() {
        let batch = vec![
            json!({"metadata": {"flagged": false}}),
            json!({"metadata": {"flagged": true}}),
            json!({"metadata": {"policy_violation": true}}),
            json!({"metadata": {}}),
        ];
        // 2 of 4 flagged -> 1.0 - 0.5 = 0.5
        assert_eq!(self_reported_compliance(&batch), 0.5);
    }

    #[test]
    fn self_reported_compliance_empty_batch_is_clean() {
        assert_eq!(self_reported_compliance(&[]), 1.0);
    }

    // --- entry_covered_entity_address ---

    #[test]
    fn covered_entity_address_read_from_first_carrying_entry() {
        let batch = vec![json!({"metadata": {}}), json!({"metadata": {"covered_entity_address": "0xabc123"}})];
        assert_eq!(entry_covered_entity_address(&batch), Some("0xabc123".to_string()));
    }

    #[test]
    fn covered_entity_address_absent_returns_none() {
        assert_eq!(entry_covered_entity_address(&[json!({"metadata": {}})]), None);
    }
}

File: integrity-sdk/integrity_sdk/telemetry/derive.py
"""
Derives the four Agent Integrity Score input signals (entropy, grounding,
sacrifice, compliance) from a batch of telemetry entries, for the SDK to
attach to a `POST /v1/telemetry/ingest` call (see
docs/INTERFACE_CONTRACT.md's telemetry ingestion section).

Ownership boundary (load-bearing, see scoring-core/src/lib.rs's own module
docstring): this file produces the AIS formula's *inputs*, never the score
itself. integrity-oracle's `scoring-core` crate remains the only place the
weighted-sum formula is computed; every function here is documented as a
first-pass heuristic the SDK can compute client-side, which the oracle may
combine with its own server-side signals (verified GPU-hours, ZK
attestation state) rather than trust blindly.

All four signals are normalized to `[0.0, 1.0]` with a consistent polarity:
**1.0 always means "best/most trustworthy"**, 0.0 always means "worst" — this
matches the AIS formula's own S_entropy/S_grounding/S_sacrifice/S_compliance
polarity (see scoring-core's `AisComponentInputs` field docs: "higher is
better" for all four axes), so a caller never has to remember that one
signal is inverted relative to the others.

Every entry in `batch` is a plain dict, one per `IntegrityClient.log_telemetry`
call — see client.py for the exact shape (`metadata`, optional pre-computed
`entropy`/`grounding`). This module doesn't depend on client.py to avoid a
circular import; client.py imports this one.
"""

from __future__ import annotations

import math
from collections import Counter
from typing import Any, Dict, List, Optional


def lexical_stability_score(text: str) -> float:
    """
    Real Shannon entropy over word-frequency distribution, normalized to
    `[0, 1]` by dividing by the maximum possible entropy for that many
    distinct words (`log2(unique_word_count)`), then INVERTED (`1 -
    normalized_entropy`) so the result reads as "how stable/consistent is
    this text" rather than "how diverse" — matching S_entropy's actual
    polarity ("rewards stability", per scoring-core's docstring).

    This is a real information-theoretic computation, not the previous
    inline heuristic (`len(set(words)) / len(words)`, a crude type-token
    ratio) that `integrations/openai_integrity.py` used to compute directly
    — that heuristic is now a call into this function, so there is exactly
    one definition of "entropy" in this SDK, not two that could silently
    drift apart.

    Empty or single-word text has no meaningful frequency distribution to
    measure entropy over, so it returns 1.0 (maximally stable) rather than
    an undefined value — a one-word response isn't erratic, there's just
    nothing to compare it against.
    """
    words = text.split()
    if len(words) <= 1:
        return 1.0

    counts = Counter(words)
    total = len(words)
    probabilities = [count / total for count in counts.values()]
    shannon_entropy = -sum(p * math.log2(p) for p in probabilities)

    max_entropy = math.log2(len(counts)) if len(counts) > 1 else 1.0
    normalized_entropy = shannon_entropy / max_entropy if max_entropy > 0 else 0.0

    return 1.0 - min(max(normalized_entropy, 0.0), 1.0)


# Deliberately crude keyword heuristic, carried over unchanged from the
# original `integrations/openai_integrity.py` implementation (just moved
# here so it has one home) — a real grounding measure would need to check
# completion claims against retrieved source documents, which requires
# integration-specific context this shared function doesn't have. Documented
# as a first-pass heuristic, not a substitute for real fact-checking.
_UNGROUNDED_MARKERS = ("hallucinate", "not sure", "i don't know", "i'm not certain")


def keyword_grounding_score(text: str) -> float:
    lowered = text.lower()
    return 0.40 if any(marker in lowered for marker in _UNGROUNDED_MARKERS) else 0.95


def _entry_entropy(entry: Dict[str, Any]) -> Optional[float]:
    """Prefers a pre-computed `entropy` field (set by an integration that
    had the completion text at hand, e.g. openai_integrity.py) over
    recomputing from `metadata.text_output` — avoids redundant work and
    lets an integration supply a better-informed value if it has one."""
    if isinstance(entry.get("entropy"), (int, float)):
        return float(entry["entropy"])
    text = entry.get("metadata", {}).get("text_output")
    if isinstance(text, str) and text:
        return lexical_stability_score(text)
    return None


def _entry_grounding(entry: Dict[str, Any]) -> Optional[float]:
    if isinstance(entry.get("grounding"), (int, float)):
        return float(entry["grounding"])
    text = entry.get("metadata", {}).get("text_output")
    if isinstance(text, str) and text:
        return keyword_grounding_score(text)
    return None


def derive_entropy(batch: List[Dict[str, Any]]) -> float:
    """Batch-mean stability score across every entry that has a completion
    text or a pre-computed value. Returns 1.0 (no evidence of instability)
    for an empty batch or a batch with no scoreable entries — an agent that
    hasn't produced any output yet shouldn't be penalized as if it had
    produced erratic output."""
    values = [v for v in (_entry_entropy(e) for e in batch) if v is not None]
    return sum(values) / len(values) if values else 1.0


def derive_grounding(batch: List[Dict[str, Any]]) -> float:
    values = [v for v in (_entry_grounding(e) for e in batch) if v is not None]
    return sum(values) / len(values) if values else 1.0


# Ceiling chosen so a single, realistically-sized agent session (a few dozen
# calls at typical chat-completion lengths) doesn't saturate the score, while
# a session doing meaningfully more real work scores higher — deliberately
# not tied to any specific model's context window, since this batch may mix
# providers/models.
_SACRIFICE_TOKEN_CEILING = 200_000


def derive_sacrifice(batch: List[Dict[str, Any]]) -> float:
    """
    Proxy for scoring-core's "costly, hard-to-fake evidence of real resource
    commitment" axis (see `AisComponentInputs.gpu_hours_verified`'s
    docstring). The SDK has no way to independently VERIFY compute
    contribution client-side (that's exactly why the oracle's field is
    named `_verified`, not `_claimed`) — total tokens actually processed
    across the batch is the closest honestly-observable proxy available at
    this layer: it's real, it's already present in both integrations'
    telemetry (`usage.prompt_tokens`/`completion_tokens` from OpenAI,
    `token_usage` from LangChain's `llm_output`), and it can't be
    cost-free-faked the way a self-reported flag could. This is a genuinely
    weaker signal than oracle-verified GPU-hours and is documented as such —
    the oracle's own ingestion handler decides how much weight to give a
    client-reported sacrifice signal versus its own server-side evidence.
    """
    total_tokens = 0
    for entry in batch:
        metadata = entry.get("metadata", {})
        usage = metadata.get("token_usage") or {}
        if isinstance(usage, dict):
            total_tokens += int(usage.get("total_tokens", 0) or 0)
            total_tokens += int(usage.get("prompt_tokens", 0) or 0)
            total_tokens += int(usage.get("completion_tokens", 0) or 0)
        for key in ("input_tokens", "output_tokens"):
            value = metadata.get(key)
            if isinstance(value, (int, float)):
                total_tokens += int(value)

    if total_tokens <= 0:
        return 0.0
    return min(math.log10(total_tokens + 1) / math.log10(_SACRIFICE_TOKEN_CEILING + 1), 1.0)


def derive_compliance(
    batch: List[Dict[str, Any]],
    *,
    compliance_gate_address: Optional[str] = None,
    covered_entity_address: Optional[str] = None,
    w3: Optional[Any] = None,
) -> float:
    """
    Combines self-reported compliance signals from the telemetry batch with
    a live on-chain `ComplianceGate.isHealthcareCompliant` read when chain
    access is available — **on-chain wins** when both are present, since a
    self-report alone is exactly the kind of unverified claim
    `ComplianceGate.sol`'s own NatSpec warns against trusting for a
    regulated-vertical agent (see contracts/src/shield/ComplianceGate.sol).

    Self-reported signal: fraction of batch entries NOT flagged as a policy
    violation (`metadata.get("policy_violation")` or `metadata.get("flagged")`
    truthy), mirroring `telemetry/conventions.py`'s
    `IntegrityAttributes.COMPLIANCE_*` attribute intent.
    """
    flagged_count = 0
    total = len(batch)
    for entry in batch:
        metadata = entry.get("metadata", {})
        if metadata.get("policy_violation") or metadata.get("flagged"):
            flagged_count += 1

    self_reported = 1.0 - (flagged_count / total) if total > 0 else 1.0

    if compliance_gate_address and covered_entity_address and w3 is not None:
        try:
            from .. import chain as chain_module

            gate_artifact = chain_module._load_artifact("ComplianceGate")
            gate = w3.eth.contract(address=w3.to_checksum_address(compliance_gate_address), abi=gate_artifact["abi"])
            is_compliant = gate.functions.isHealthcareCompliant(
                w3.to_checksum_address(covered_entity_address)
            ).call()
            # On-chain wins: a live "not compliant" read overrides a clean
            # self-report (an agent can't talk its way out of a lapsed BAA),
            # but a live "compliant" read still can't push the score above
            # what self-reporting already earned — on-chain presence proves
            # eligibility, not good behavior within that eligibility.
            return min(self_reported, 1.0) if is_compliant else 0.0
        except Exception:
            # Chain read failures (RPC down, gate not deployed, etc) fall
            # back to the self-reported signal rather than raising — this
            # function computes an input to a score, not a security gate;
            # EHRGate.sol remains the real, fail-closed enforcement point
            # for actual PHI access, this is just a reputation input.
            pass

    return self_reported


# Decimal places every derived signal is quantized to before it is signed.
#
# This is a CORRECTNESS requirement of the signature scheme, not cosmetic
# rounding (FIXED 2026-07-17 — was silently rejecting ~20% of real, correctly
# signed telemetry with a 400).
#
# `client.flush_telemetry` signs the canonical JSON of a payload containing
# these floats, and integrity-oracle re-serializes the same payload with
# Rust's `serde_json` to check that signature. Both sides emit the "shortest
# string that round-trips back to this exact f64" — but when a float has TWO
# equally-short round-tripping representations, Python's repr (David Gay) and
# Rust's ryu are each free to pick a different one. Nothing is wrong with
# either; they simply disagree, the canonical bytes differ, and Ed25519
# verification fails on a payload that was signed perfectly correctly.
#
# Confirmed against the live oracle, not theorised: a real derived entropy of
# 0.011890908425879365 failed every time while 0.009712883245855508 passed,
# and in Python BOTH "0.011890908425879365" and "0.011890908425879366"
# round-trip to that identical f64 (hex 0x1.85a42b6789780p-7) — the exact
# two-candidate ambiguity above. The oracle's error surfaced as a confusing
# "eip191 verification error: signature must be 65 bytes, got 64", which is a
# downstream red herring: `crypto::verify_agent_signature` tries Ed25519
# first, gets `false` (not an error), and falls through to the EIP-191 branch,
# which then chokes on a 64-byte Ed25519 signature.
#
# The ambiguity is a ~17-significant-digit phenomenon; at 6 decimal places the
# shortest round-tripping representation is unique, so both languages
# necessarily agree. 6dp is also far more precision than these heuristics
# justify (see each derive_* docstring — they are first-pass client-side
# estimates the oracle independently recomputes anyway), so nothing of value
# is lost by quantizing.
#
# NOTE (real remaining gap, deliberately not papered over): this only fixes the
# floats the SDK itself generates. A caller passing an arbitrary float through
# `log_telemetry(metadata=...)` can still land on an ambiguous value and hit
# the same rejection, since that value is signed verbatim inside `otel_spans`.
# The general fix is a shared canonicalization standard with a fully specified
# number format on both sides -- RFC 8785 (JCS) mandates ECMAScript's
# Number::toString, which is deterministic -- rather than each language's own
# shortest-repr. Flagged in PRODUCTION_GAPS.md rather than silently assumed
# away, same as bcc.py's own canonicalization docstring does for a related
# non-ASCII concern.
_SIGNAL_DECIMALS = 6


def derive_ais_signals(
    batch: List[Dict[str, Any]],
    *,
    compliance_gate_address: Optional[str] = None,
    covered_entity_address: Optional[str] = None,
    w3: Optional[Any] = None,
) -> Dict[str, float]:
    """Bundles all four derived signals into the shape
    `POST /v1/telemetry/ingest`'s `derived_signals` field expects (see
    docs/INTERFACE_CONTRACT.md).

    Every value is quantized to `_SIGNAL_DECIMALS` decimal places — see that
    constant's comment for why this is load-bearing for signature
    verification and not a cosmetic choice.
    """
    return {
        "entropy": round(derive_entropy(batch), _SIGNAL_DECIMALS),
        "grounding": round(derive_grounding(batch), _SIGNAL_DECIMALS),
        "sacrifice": round(derive_sacrifice(batch), _SIGNAL_DECIMALS),
        "compliance": round(
            derive_compliance(
                batch,
                compliance_gate_address=compliance_gate_address,
                covered_entity_address=covered_entity_address,
                w3=w3,
            ),
            _SIGNAL_DECIMALS,
        ),
    }
